Getting 401 when Community or Collection Administrator tries to edit Collection Authorization Group

Raf-atmire opened the pull request:
## References
* Backend changes to support https://github.com/DSpace/dspace-angular/issues/927  (A complete fix also requires Angular PR which is in progress)

## Description
This PR provides us with 3 new feature calls to check if:
- A given EPerson is a community admin of any community -> CommunityAdminFeature
- A given EPerson is a collection admin of any collection -> CollectionAdminFeature
- A given EPerson is a community or collection admin of any community or collection -> ComColAdminFeature

This PR also provides us with a check if someone can manage groups for a community or collection admin without iterating all communities and collections in either Angular or REST -> ManageGroupsFeature

It also provides us with two new endpoints that'll query our SOLR to find Communities and Collections respectively to find which ones the current user is an Admin for. findAdminAuthorized on both the CommunityRestRepository and CollectionRestRepository respectively.

## Instructions for Reviewers

List of changes in this PR:
* Added the above mentioned three features to determine admin permissions on community/collection for a given user
* Added the feature to determine whether or not a given user can manage groups
* Added the two endpoints that'll allow us to search communities and collections for which a given user is an admin

This PR can be tested by creating an EPerson that is an admin, an EPerson that is a collection admin, an EPerson that's a Community admin and just a regular user. You can then proceed by testing each feature individually to verify that you'll get the expected result for each logged in user. Please refer back to @benbosman 's proposal for a detailed rundown: https://github.com/DSpace/dspace-angular/issues/927#issuecomment-758604998

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

peter-atmire committed:
76284: Issue 927: 401 Community or Collection Administrator editing Collection Authorization Group REST Implementation

- implement Community and Collection admin search
- implement the CollectionAdminFeature and CommunityAdminFeature Authorization feature

Raf-atmire committed:
Small refactor

peter-atmire committed:
76284: Feedback

Raf-atmire committed:
Merge remote-tracking branch 'dspace/main' into w2p-76284_issue-927---non-site-admin-edit-collection-authorization-group-REST

Conflicts:
	dspace-api/src/main/java/org/dspace/discovery/SolrServiceResourceRestrictionPlugin.java

Raf-atmire committed:
Merge branch 'w2p-76284_issue-927---non-site-admin-edit-collection-authorization-group-REST' of https://github.com/atmire/DSpace into w2p-76284_issue-927---non-site-admin-edit-collection-authorization-group-REST

Raf-atmire committed:
Merge remote-tracking branch 'dspace/main' into w2p-76284_issue-927---non-site-admin-edit-collection-authorization-group-REST

Conflicts:
	dspace-server-webapp/src/test/java/org/dspace/app/rest/CollectionRestRepositoryIT.java

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/ManageGroupsFeatureIT.java:
```diff
@@ -0,0 +1,1687 @@
+            .andExpect(status().isOk())
+            .andExpect(
+                jsonPath("$._embedded.authorizations[?(@._embedded.feature.id=='canManageGroups')]")
+                    .exists());
```
Same thing here...if you've *disabled* the ability of Community Admins to manage these groups, shouldn't this return `doesNotExist()`?  (This question applies to all of the `*NoCollectionGroupPermission` tests)

Raf-atmire replied:
Please refer to: https://github.com/DSpace/DSpace/pull/3148#discussion_r580184087

tdonohue reviewed (changes requested):
@Raf-atmire : I gave this a code review today.  Mostly it looks good, but some tests seem **incorrect** to me, which makes me worried there may be a bug in this code.  See inline comments for more details.  I haven't tested this yet manually.

Raf-atmire commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/ManageGroupsFeatureIT.java:
```diff
@@ -0,0 +1,1687 @@
+            .andExpect(status().isOk())
+            .andExpect(
+                jsonPath("$._embedded.authorizations[?(@._embedded.feature.id=='canManageGroups')]")
+                    .exists());
```
We've indeed disabled the Community Admin's permissions, but the Collection Admin's permissions are still set to "true". Which means that, indeed, community groups won't be able to be managed, but the collection ones are still able to be managed (all the configs are true by default). This feature determines if the user can manage any of those groups, not just one set in particular.
The tests that solidify the use-case of not being able to manage any, can be found here and below: https://github.com/DSpace/DSpace/pull/3148/files#diff-875d18cdd848890a5b752290bf27da005f24829dd48a11b0bc113a96ddb25568R1199
Keep in mind that a general admin should always be able to manage groups

tdonohue replied:
@Raf-atmire : Thanks for the explanation. That makes sense now...I obviously didn't realize it when reviewing these tests.

Raf-atmire committed:
Refactored ManageGroupFeature functionality

peter-atmire committed:
76284: Add graphical community/collection structure

Raf-atmire commented:
@tdonohue Thank you for the feedback! I think we've addressed all of it and created a PR for the changes to the RESTContract: https://github.com/DSpace/RestContract/pull/157
Could you have another look when you get the chance? Thanks!

tdonohue reviewed (changes requested):
@Raf-atmire : Thanks for the updates & for answering my prior concerns.  I've re-reviewed & tested this.  It works overall & looks correct.  However, I have a few minor things that need fixed up (currently the new search endpoints are not checking that you are authenticated).

Beyond that, honestly this looks good and I'd be +1 once these last few things are fixed.

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/CollectionRestRepositoryIT.java:
```diff
@@ -2357,4 +2375,545 @@ public void findOneTestWithEmbedsWithInvalidPageSize() throws Exception {
+
+    @Test
+    public void testAdminAuthorizedSearch() throws Exception {
+        setUpAuthorizedSearch();
```
can we replace this explicit call and the other with a @Before annotation over the setUpAuthorizedSearch method?

Raf-atmire replied:
We can't really use the @Before annotation here since there are a multitude of tests inside this IT class that do not require or want this setupAuthorizedSearch to happen, it'd result in other tests being affected when they shouldn't be

abollini replied:
thanks for explaining

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityRestRepositoryIT.java:
```diff
@@ -1934,4 +1956,491 @@ public void createTestInvalidParentCommunityBadRequest() throws Exception {
+
+    @Test
+    public void testAdminAuthorizedSearch() throws Exception {
+        setUpAuthorizedSearch();
```
can we use the Before annotation for that?

abollini reviewed (changes requested):
thanks @Raf-atmire for this PR. I would like to make a second review once these first comments have been addressed

peter-atmire committed:
76284: non site admin edit collection authorization group REST feedback

- Merge AuthorizeSolrService into AuthorizeService
- PreAuthorize findAdminAuthorized methods

Raf-atmire committed:
Changed to primitive long for count methods in authorizeService

peter-atmire committed:
76284: Use primitive longs

abollini commented on dspace-api/src/main/java/org/dspace/authorize/AuthorizeServiceImpl.java:
```diff
@@ -787,4 +802,186 @@ public ResourcePolicy createOrModifyPolicy(ResourcePolicy policy, Context contex
+     *                  false when this is not the case, or an exception occurred
+     */
+    @Override
+    public boolean isCommunityAdminSolr(Context context) throws SQLException {
```
it would be better to have a single method isCommunityAdmin that only check via SOLR. The code actually used for the database lookup should be moved in the indexer plugin that is mandatory to be used to have such implementation working properly. Having two different implementations is confusing and would lead to inefficiency as we could use the "old" one instead than the new one in most of the code base.

If you feel not comfortable with such change I suggest to create a code task for future improvement

tdonohue replied:
@Raf-atmire : I agree with @abollini here.  Would it be possible to simply *replace* the current `isCommunityAdmin()` methods with the contents of this new one?  I'd rather there be a single `isCommunityAdmin()` method which queries Solr, than have two different options.

Also, it's worth noting that the `isCommunityAdmin()` method which accesses the database seems to be primarily used by the REST API.  So, currently, parts of the REST API are using `isCommunityAdmin()` while other parts use `isCommunityAdminSolr()`.  This could result in odd behavior if the two are not identical.

If there is some reason why you chose this approach (of having two implementations), I'd be OK with moving this to a follow-up ticket.  But, I wanted to note that I don't like have two methods here doing the same thing in different ways.

abollini commented on dspace-api/src/main/java/org/dspace/authorize/AuthorizeServiceImpl.java:
```diff
@@ -787,4 +802,186 @@ public ResourcePolicy createOrModifyPolicy(ResourcePolicy policy, Context contex
+     *                  false when this is not the case, or an exception occurred
+     */
+    @Override
+    public boolean isCollectionAdminSolr(Context context) throws SQLException {
```
as for isCommunityAdminSolr

tdonohue replied:
Agreed here as well...it's the same issue.  It'd be nice if the Solr code was moved directly into `isCollectionAdmin()` and this method was removed.  Again, though, if there's a reason for two implementations of `isCollectionAdmin` (one for DB and one for Solr), I'd be OK with moving this to a followup ticket

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java:
```diff
@@ -200,6 +200,22 @@ public CollectionRest findOne(Context context, UUID id) {
+            List<Collection> collections = authorizeService.findAdminAuthorizedCollection(context, query,
+                Math.toIntExact(pageable.getOffset()),
+                Math.toIntExact(pageable.getPageSize()));
+            Long tot = authorizeService.countAdminAuthorizedCollection(context, query);
```
https://lgtm.com/projects/g/DSpace/DSpace/rev/pr-bc08c3f246f93557c9f6f0bb642b09902a75771e

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java:
```diff
@@ -203,6 +204,22 @@ public CommunityRest findOne(Context context, UUID id) {
+            List<Community> communities = authorizeService.findAdminAuthorizedCommunity(context, query,
+                Math.toIntExact(pageable.getOffset()),
+                Math.toIntExact(pageable.getPageSize()));
+            Long tot = authorizeService.countAdminAuthorizedCommunity(context, query);
```
https://lgtm.com/projects/g/DSpace/DSpace/rev/pr-bc08c3f246f93557c9f6f0bb642b09902a75771e

abollini reviewed (commented):
I leave the decision about the merge to @tdonohue I have added some suggestions inline but none of them are blocking

tdonohue reviewed (commented):
@Raf-atmire : Like @abollini , I have questions about the new duplicative `isCommunityAdminSolr()` and `isCollectionAdminSolr()` methods.  I was expecting you to just change `isCommunityAdmin()` and `isCollectionAdmin()` to use Solr instead of the DB.  Is there a reason you chose this other direction?

Overall, this PR looks good & it works. I'd just like more info on these duplicative methods so that we can make a decision on whether to replace them now, or move this into follow-up work.  Thanks!

Raf-atmire committed:
Partial refactor, still issues with infinite loops

Raf-atmire commented:
Hi @tdonohue @abollini , whilst I certainly agree with the comments about making one method instead of having 2 separate methods, we've ran into some issues.

We've currently hit a point with the development where we keep running into the same issues, repeating itself across a multitude of different places in DSpace.
When trying to combine the `isCommunityAdmin` functionalities and enabling it to be just a solr call instead of a database call, we're running into the issue that that particular Solr call calls forth the plugin framework. This plugin framework will then try to add additional filterqueries or logic to the DiscoverQuery, which is fine in its core. The issue arises from those specific plugins calling the exact same `isCommunityAdmin` call to determine whether something should be added to the DiscoverQuery. This results in an infinite loop. This also underlines performances issues that for each single solr query, a call is made to the database (right now) to check if the user is a community/collection admin. This call is particularly slow for users with a lot of groups or repositories with a lot of communities and collections.
We'd need to somehow ensure that those calls aren't being made if it's not necessary. For example, some plugins don't make sense to be ran for communities and collections since they're item specific, however currently they do get processed for each and every Solr request. 
If we were to limit the Plugin framework to only run X plugin for Y DSpaceObject, we could already tone down the issue by ensuring that the `isCommunityAdmin` calls aren't being made for a DSO for which it doesn't make sense. However, this would still mean that we could potentially run into the infinite loop problemn again.
The second possible solution is to create a framework in which the Plugins are activatable for a specific DiscoveryConfiguration. This would mean that, for certain Solr calls, we could use a DiscoveryConfiguration that does not have all the Plugins, but only the ones it really needs. Whereas the default plugin could have all the plugins enabled by default. This would allow us to ensure that the `isCommunityAdmin` calls do not use the plugins that it doesn't need and that it doesn't produce an infinite loop that way.
Regardless of the solution that we chose to go for, this needs to be a community decision and a subject that should be independent of this ticket since a major design change will need to happen.

I've started working on this in this branch: https://github.com/atmire/DSpace/tree/w2p-76284_issue-927-refactoring-authorize-methods
This branch still contains the infinite loop issue through this line of code: https://github.com/atmire/DSpace/blob/w2p-76284_issue-927-refactoring-authorize-methods/dspace-api/src/main/java/org/dspace/discovery/SolrServiceResourceRestrictionPlugin.java#L163

mwoodiupui commented:
So, uh, why are we going to Solr for this, if it is troublesome?

tdonohue commented:
@Raf-atmire : Thanks for the thorough summary.  I was talking briefly with @benbosman over Slack, and came to the realization that the line of code you reference doesn't seem to be needed in `SolrServiceResourceRestrictionPlugin`.  If you look closely, there's duplicative logic in that plugin & it'd be cleaner to *remove* the calls to `isCommunityAdmin` and `isCollectionAdmin`:

1. First, as you noted, currently, it has an `if` which checks if you are either a Community or Collection Admin: https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/discovery/SolrServiceResourceRestrictionPlugin.java#L162-L163
2. Looking at the logic of the current `isCommunityAdmin()` method, it's just looking up all your Groups & checking corresponding ResourcePolicies for ADMIN on a COMMUNITY: https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/authorize/AuthorizeServiceImpl.java#L439-L441  (`isCollectionAdmin()` acts similarly)
3. However, if you look at what is *inside* that `if`, you'll see it then calls `createLocationQueryForAdministrableItems`, which again looks up all your Groups, and checks for ResourcePolicies for ADMIN either on COMMUNITY or COLLECTION: https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java#L552-L563

So, steps 2 and 3 are doing the same general thing twice. 

Assuming this is the only remaining issue with migrating us over to using Solr for `isCommunityAdmin()` and `isCollectionAdmin()`, then it should be relatively easy to solve:

1. Refactor `SolrServiceResourceRestrictionPlugin.addtionalSearchParameters()` to **remove** that `if` statement & the calls to `isCommunityAdmin()` and `isCollectionAdmin()`.
2. Change the code instead to immediately call `createLocationQueryForAdministrableItems(context)` (as this does the same check).  If that method returns a non-empty String, then we append it to the query (with an OR before it), as we know this person has Community or Collection admin rights.  If, instead, that method returns an empty String (or null), do nothing.
3. This likely will require a small refactor to `createLocationQueryForAdministrableItems` to [remove this `log.warn`](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java#L605-L606), as we are no longer concerned if that method returns an empty string or null.

All that said, as I told @benbosman , if you find this is **not** the only major issue with swapping us over to Solr implementations, then I'm OK with leaving this PR "as-is" and logging a ticket to refactor this at a later date.   It just looks like, at a glance, your current branch has solved most of the issues already...and this might be the only fix necessary for a full solution.

(@mwoodiupui : Just a sidenote to answer your question, we've switched to Solr here because the Database implementations are not performing well...if you look at the underlying code, it's a complex query that also relies heavily on the Group2GroupCache (as you need to query all resource policies for all groups you are in, including parent groups of those groups, etc.)  We've determined it's simply a lot faster to index this information (using that same complex logic) and grab it out of Solr much more quickly whenever it is needed.)

Raf-atmire committed:
fix tests

Raf-atmire committed:
Merge branch 'w2p-76284_issue-927---non-site-admin-edit-collection-authorization-group-REST' of https://github.com/atmire/DSpace into w2p-76284_issue-927---non-site-admin-edit-collection-authorization-group-REST

Raf-atmire committed:
Merge remote-tracking branch 'dspace/main' into w2p-76284_issue-927---non-site-admin-edit-collection-authorization-group-REST

Raf-atmire committed:
fix duplicate id

tdonohue commented on dspace-api/src/test/java/org/dspace/authorize/AuthorizeServiceTest.java:
```diff
@@ -129,70 +127,72 @@ public void testauthorizeMethodRespectSpecialGroups() {
-        }
-    }
+//
+//    @Test
```
@Raf-atmire : It looks like these tests were accidentally left commented out?  Was that purposeful?  If so, let's create a ticket to follow this PR up with a fix for these tests (if for some reason they currently don't work).

tdonohue replied:
I tried reenabling them, and sure enough they don't work right now, as they seem to assume the Database approach instead of the new Solr approach.  I will treat this as a *follow-up* task, rather than requiring it in this PR.  The new Solr approach to determining Community/Collection Admins already has ITs that prove it works. So, we can refactor these Unit Tests in a follow-up ticket.

tdonohue reviewed (approved):
👍 Looks great, thanks @Raf-atmire!  I've tested this with the corresponding Angular PR and it all works perfectly.

For the minor issues of the commented out unit tests, I've logged a follow-up ticket which is scheduled for 7.0 final: #3216
